### ISSUE TITLE
Provide availability guarantees for IBC Packet data

### ISSUE THREAD

**Issue by ethanfrey** [cosmos/ibc-go#74]
## Summary

Although the actual packet data is not required to be accessible in the MultiStore of the sending chain for the application logic to function, it requires extremely high availability guarantees for IBC to be reliable. This information is currently only available via the event system, which has numerous issues and cannot provide such guarantees. 

Thus, I propose to add the original (unhashed) packet and acknowledgement data to the MultiStore to ensure a reliable IBC implementation. They should also be cleaned up at the same time that the commitments (hashed values) are.

This is requested to be included for 0.42.0 and I would like a timely review and discussion.

## Problem Definition

Much of this came to a surprise to me while working on IBC integration tests, and now the [ts-relayer](https://github.com/confio/ts-relayer). Until recently, I thought I just didn't fully understand the dozens of gRPC endpoints, but then discovered that [on SendPacket, the packet data is never written to the store](https://github.com/cosmos/cosmos-sdk/blob/17d7e9a8d51e11340aab1705a3d47b6f98abff6f/x/ibc/core/04-channel/keeper/packet.go#L112-L116). But rather, [only written to the event system](https://github.com/cosmos/cosmos-sdk/blob/17d7e9a8d51e11340aab1705a3d47b6f98abff6f/x/ibc/core/04-channel/keeper/packet.go#L121-L134).

The event system has been woefully unmaintained for years and there is now a [tracking issue in Tendermint](https://github.com/tendermint/tendermint/issues/5963) to start working on an overhaul. In particular, there are two main issues I highlighted that directly effect the suitability for IBC:

* https://github.com/tendermint/tendermint/issues/3665 (reported May 2019) is that events emitted in BeginBlock and EndBlock are not searchable. In fact, they are not written anywhere afaik (you cannot get a list of BeginBlock events when querying a block by height).
* https://github.com/tendermint/tendermint/issues/5133 is that multiple events of the same type in the same transaction (even different messages) get merged into one.

Before this is dismissed as purely theoretical complaints, I want to say that Band protocol [hit problems in integration as they send packets in EndBlock](https://discord.com/channels/669268347736686612/723170066937413693/808611836954017792), which is blocked by the first issue. While working on the relayer, I figured sending multiple ReceivePacket messages in one tx is a nice way to include them in one block without worrying about sequence issues and tx ordering. Except that the relayer wouldn't be able to parse the acknowledgement data from the resulting events due to the second issue. I am sure there are more cases (and will be many more in the coming months of IBC development), but I want to highlight that this has become a blocker *even before IBC is live on the Cosmos Hub*

While there is some [motion to upgrade the event system in Tendermint](https://github.com/tendermint/tendermint/discussions/6114) this will not happen overnight, and we do need a reliable solution for sending IBC packets very soon. 

## Proposal

Store the data:

To `ibc/core/04-channel/keeper/Keeper` add `SetPacketData(ctx, data)` and `SetAcknowledgementData(ctx, data)`. Both will store the data in a lookup by the hashed value (commitment), which is already stored in the system. In `SendPacket`, call `SetPacketData()` just after [`SetPacketCommitment`](https://github.com/cosmos/cosmos-sdk/blob/17d7e9a8d51e11340aab1705a3d47b6f98abff6f/x/ibc/core/04-channel/keeper/packet.go#L116). In `WriteAcknowledgement`, call 
`SetAcknowledgementData()` just after [`SetPacketAcknowledgement`](https://github.com/cosmos/cosmos-sdk/blob/17d7e9a8d51e11340aab1705a3d47b6f98abff6f/x/ibc/core/04-channel/keeper/packet.go#L358-L361).

Clean up the data:

When deleting the packet commitment: [here](https://github.com/cosmos/cosmos-sdk/blob/17d7e9a8d51e11340aab1705a3d47b6f98abff6f/x/ibc/core/04-channel/keeper/packet.go#L500) and [here](https://github.com/cosmos/cosmos-sdk/blob/17d7e9a8d51e11340aab1705a3d47b6f98abff6f/x/ibc/core/04-channel/keeper/timeout.go#L144), ensure we also delete the PacketData at the same time. NB: it seems that acknowledgement commitments are currently never deleted, but cleanup of the data should happen when the commitment is removed.

Expose gRPC endpoints:

The current gRPC endpoints allows us to get the commitments for a channel. Add two more endpoints to QueryPacketByCommitment and QueryAcknowledgementByCommitment to look up the data that we wrote above. Once this is working, it becomes usable by the clients. Adding more gRPC endpoints to make the queries simpler can be done in the future *in a non-breaking manner*.

## Backwards compatibility

I write this assuming that it is included in v0.42.0. If not, the issue may become more difficult as people start working on custom work-arounds to enable their chain's use cases.

While the Cosmos Hub will be upgraded to v0.41.0 of the Cosmos-SDK, it is highly unlikely that other mainnets will use this, as there is no support for slashing on double-signing if they don't use the "cosmos" bech32 prefix. cosmos/cosmos-sdk#8461 So, we can assume that all other mainnets will be based on Cosmos SDK v0.42.0 or higher, and the Cosmos Hub will likely upgrade when it adds more advanced functionality.

In order to maintain compatibility, I propose that we continue to emit the same events in v0.42.0, but include these new endpoints *in addition*. All existing relayers will continue to work without problems and when they wish, they can add an optional flag (per chain) to use the new gRPC endpoints instead of the event system for these packet queries.

An upgrade from v0.41 -> v0.42 would add this new storage and index all packets send *after the upgrade*. The fact that previous packets are not indexed there is not an issue for the relayer (but must be handled gracefully by the cleanup code). The relayer can continue to run in "legacy" event mode for some days after an upgrade, before switching on the new "use gRPC" flag.

Review requested by @cwgoes @ebuchman @aaronc (and @clevinson for 0.42 release planning)

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by crodriguezvega** [comment]
@ethanfrey Is this issue still relevant now that issue [#5963](https://github.com/tendermint/tendermint/issues/5963) in Tendermint is completed?

**Comment by adizere** [comment]
The indexer in Tendermint is undergoing some reparations again (https://github.com/tendermint/tendermint/issues/9712). The Tendermint team plans to overhaul storage and indexing over the next quarters, eg. see https://github.com/tendermint/tendermint/pull/9437 for great discussion. I suspect this issue can be closed and we can following up in Tendermint issues/PRs.
